File: library/network/src/main/java/com/example/network/utils/GsonUtil.java
package com.example.network.utils;

import android.text.TextUtils;

import com.google.gson.Gson;
import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonParser;
import com.google.gson.reflect.TypeToken;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/**
 * gson解析工具类
 *
 * @author yuanwai
 */
public class GsonUtil {

    private static Gson gson;

    public static Gson getGson(){
        if (gson == null) {
            synchronized (Gson.class) {
                if (gson == null) {
                    gson = new Gson();
                }
            }
        }
        return gson;
    }

    /**
     * 将json转换为bean
     */
    public static <T> T getBean(String json, Class<T> cls) {
        if (TextUtils.isEmpty(json)) return null;
        try {
            return getGson().fromJson(json, cls);
        } catch (Exception e) {
            e.printStackTrace();
            return null;
        }
    }

    /**
     * 将json转换为bean或list
     */
    public static Object getObjectOrList(String json, Class<?> clazz) {
        try {
            JsonParser jsonParser = new JsonParser();
            JsonElement jsonElement = jsonParser.parse(json);
            if (jsonElement.isJsonArray()) {
                JsonArray jsonArray = jsonElement.getAsJsonArray();
                List<Object> list = new ArrayList<Object>();
                for (JsonElement element : jsonArray) {
                    if (element.isJsonArray()) {
                        list.add(getObjectOrList(element.toString(), clazz));
                    } else {
                        list.add(getGson().fromJson(element, clazz));
                    }
                }
                return list;
            } else {
                return getGson().fromJson(json, clazz);
            }
        } catch (Exception e) {
            return null;
        }
    }

    /**
     * 将json转换为string
     */
    public static String GsonToString(Object object) {
        if (object == null) {
            return "";
        }
        try {
            return getGson().toJson(object);
        } catch (Exception e) {
            e.printStackTrace();
            return "";
        }
    }

    /**
     * 将json转换为Map<String, Object>
     */
    public static Map<String, Object> getMap(String jsonString) {
        try {
            return getGson().fromJson(jsonString,
                    new TypeToken<Map<String, Object>>() {
                    }.getType());
        } catch (Exception e) {
            e.printStackTrace();
            return null;
        }
    }

}

File: library/network/src/main/java/com/example/network/utils/GsonUtil.kt
package com.example.network.utils

import android.text.TextUtils
import com.google.gson.Gson
import com.google.gson.JsonParser
import com.google.gson.reflect.TypeToken
import java.util.*

/**
 * gson解析工具类
 *
 * @author yuanwai
 */
object GsonUtil {
    var gson: Gson? = null
        get() {
            if (field == null) {
                synchronized(Gson::class.java) {
                    if (field == null) {
                        field = Gson()
                    }
                }
            }
            return field
        }
        private set

    /**
     * 将json转换为bean
     */
    fun <T> getBean(json: String?, cls: Class<T>?): T? {
        return if (TextUtils.isEmpty(json)) null else try {
            gson!!.fromJson(json, cls)
        } catch (e: Exception) {
            e.printStackTrace()
            null
        }
    }

    /**
     * 将json转换为bean或list
     */
    fun getObjectOrList(json: String?, clazz: Class<*>?): Any? {
        return try {
            val jsonParser = JsonParser()
            val jsonElement = jsonParser.parse(json)
            if (jsonElement.isJsonArray) {
                val jsonArray = jsonElement.asJsonArray
                val list: MutableList<Any?> =
                    ArrayList()
                for (element in jsonArray) {
                    if (element.isJsonArray) {
                        list.add(getObjectOrList(element.toString(), clazz))
                    } else {
                        list.add(gson!!.fromJson(element, clazz))
                    }
                }
                list
            } else {
                gson!!.fromJson(json, clazz)
            }
        } catch (e: Exception) {
            null
        }
    }

    /**
     * 将json转换为string
     */
    @JvmStatic
    fun GsonToString(`object`: Any?): String {
        return if (`object` == null) {
            ""
        } else try {
            gson!!.toJson(`object`)
        } catch (e: Exception) {
            e.printStackTrace()
            ""
        }
    }

    /**
     * 将json转换为Map<String></String>, Object>
     */
    fun getMap(jsonString: String?): Map<String, Any>? {
        return try {
            gson!!
                .fromJson<Map<String, Any>>(
                    jsonString,
                    object :
                        TypeToken<Map<String?, Any?>?>() {}.type
                )
        } catch (e: Exception) {
            e.printStackTrace()
            null
        }
    }
}